File: Course5_BackEnd/ch01_kotlin/java-example/src/main/java/org/example/ex10/Exam10.java
package org.example.ex10;

import java.util.Optional;

public class Exam10 {
    public Exam10(ExamUser examUser) {
        Optional<ExamUser> optionalUser = Optional.ofNullable(examUser);
        optionalUser.ifPresent(it -> {
            Optional.ofNullable(examUser.getName()).ifPresent(name -> {
                if (!name.isBlank()) {
                    System.out.println(name);
                }
            });
        });

        // 더 직관적
        if (examUser != null && examUser.getName() != null) {
            if (StringUtils.notBlank(examUser.getName())) {
                System.out.println(examUser.getName());
            }
        }
    }

    public static void main(String[] args) {
        ExamUser user = new ExamUser();
        user.setName("abcd");
        new Exam10(user);
    }
}

class StringUtils {
    public static boolean notBlank(String value) {
        return !value.isBlank();
    }
}

class ExamUser {
    private String name;

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }
}

File: Course5_BackEnd(kotlin_spring)/ch01_kotlin/kotlin-example/src/main/kotlin/ex10/Exam10.kt
package ex10

fun main() {
    var user = ExamUser(name = "abcd")
    exam10(user)

}

fun exam10(examUser: ExamUser?) {
    // Object == Any
    if (examUser.isNotNull() && examUser?.name.isNotNullOrBlank()) {
        println(examUser?.name)

    }

    /*
    examUser?.let {
        if (it.name.isNotNullOrBlank()) {
            println(it.name)
        }
    }

    examUser?.let {
        it.name?.let {name ->
            println(name)
        }
    }?: run {
        // null
    }
    */
}

data class ExamUser(
    var name: String? = null
)

fun String?.isNotNullOrBlank(): Boolean { // 확장 함수(extensions function)
    return !this.isNullOrBlank()
}

fun Any?.isNotNull(): Boolean {
    return this != null
}